llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -1146,10 +1146,23 @@ bool AMDGPUDAGToDAGISel::isDSOffset2Legal(SDValue Base, unsigned Offset0,
   return CurDAG->SignBitIsZero(Base);
 }
 
-bool AMDGPUDAGToDAGISel::isFlatScratchBaseLegal(SDValue Base,
+bool AMDGPUDAGToDAGISel::isFlatScratchBaseLegal(SDValue Addr, SDValue Base,
```

Comment:
I don't think you can implement this in `isFlatScratchBaseLegal`. If the address is `saddr+ioffset` or `vaddr+ioffset` then this function is only called on the `saddr` or `vaddr` part, so this is not the right place to check for `nuw` addition.

Reply:
I think the whole address will be passed to the `Addr` instead of base in the case `saddr+ioffset` and `vaddr+ioffset`? see `AMDGPUDAGToDAGISel::SelectFlatOffsetImpl()` and `AMDGPUDAGToDAGISel::SelectScratchSAddr()`. I am trying to add some comment locally explaining the function:
```
// This is used to check whether the address of scratch_load/store in the
// form of `base + immediate offset` is legal with respect to the hardware's
// requirement that the SGPR/VGPR address offset in the flat scratch instruction
// should be unsigned.
```

Reply:
BTW, the function was only passed the `base` part, I have added the whole address as well as the base.

Reply:
OK, I see. I think this patch is probably OK as-is but the code that it touches could really benefit from some refactoring. For example isFlatScratchBaseLegal no longer just checks the "base" part of the address, and it has to look for ADD nodes that have already been discovered by its callers. In the case of SGPR+VGPR+offset addressing, we could check that both ADDs have the nuw flag set.

Maybe `selectScratchOffset`, `selectScratchSAddr` and `selectScratchSVAddr` could be combined into one function with flags saying whether it should match saddr or vaddr or both.

Reply:
I would argue that the job of the isFlatScratchBaseLegal() does not change too much. Its job is still to prove that base part of the address can be put in SGPR/VGPR, but with additional context, we can now do things better. So far, the extra checking happened inside it is trivial. Maybe there should be a better name for it:)

I feel we might need more code to check for both ADDs in case of SGPR+VGPR+Imm, because passing in two bases does not mean that the SGPR and VGPR are in two different ADDs. So I just leave a TODO in case someone came up with a simple code implementation later.

I am not sure what the ideal form of refactor would be for the three select* functions. I would really prefer someone else could help on that later as a separate NFC change.

Reply:
I have added the check for two ADDs, and refine the code to be more reasonable. Please help take a look. Thanks!